plutus-contract/src/Language/Plutus/Contract/Transaction.hs

```diff
@@ -0,0 +1,161 @@
+{-# LANGUAGE DeriveAnyClass         #-}
+{-# LANGUAGE DeriveGeneric          #-}
+{-# LANGUAGE DerivingStrategies     #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE TemplateHaskell        #-}
+module Language.Plutus.Contract.Transaction(
+      UnbalancedTx
+    , emptyTx
+    , inputs
+    , outputs
+    , forge
+    , requiredSignatures
+    , validityRange
+    , mergeWith
+    , toLedgerTx
+    -- * Constructing transactions
+    , unbalancedTx
+    , payToScript
+    , collectFromScript
+    , collectFromScriptFilter
+    ) where
+
+import           Control.Lens      (at, (^.))
+import qualified Control.Lens.TH   as Lens.TH
+import qualified Data.Aeson        as Aeson
+import qualified Data.Map          as Map
+import           Data.Maybe        (fromMaybe)
+import qualified Data.Set          as Set
+import           GHC.Generics      (Generic)
+
+import           Ledger            (Address, DataScript, PubKey, RedeemerScript, TxOut, TxOutRef, ValidatorScript)
+import qualified Ledger            as L
+import           Ledger.AddressMap (AddressMap)
+import qualified Ledger.Interval   as I
+import           Ledger.Slot       (SlotRange)
+import qualified Ledger.Tx         as Tx
+import           Ledger.Value      as V
+
+-- | An unsigned and potentially unbalanced transaction, as produced by
+--   a contract endpoint. See note [Unbalanced transactions]
+data UnbalancedTx = UnbalancedTx
+        { _inputs             :: [L.TxIn]
+        , _outputs            :: [L.TxOut]
+        , _forge              :: V.Value
+        , _requiredSignatures :: [PubKey]
+        , _validityRange      :: SlotRange
+        }
+        deriving stock (Eq, Show, Generic)
+        deriving anyclass (Aeson.FromJSON, Aeson.ToJSON)
+
+Lens.TH.makeLenses ''UnbalancedTx
+
+-- | An unbalanced transaction wiht no inputs and outputs, and an unbounded
+--   validity range.
+--
+emptyTx :: UnbalancedTx
+emptyTx = UnbalancedTx [] [] mempty [] I.always
+
+-- | The ledger transaction of the 'UnbalancedTx'. Note that the result
+--   does not have any signatures, and is potentially unbalanced (ie. invalid).
+--   To produce a balanced 'Tx', use 'Language.Plutus.Contract.Wallet.balance'.
+toLedgerTx :: UnbalancedTx -> L.Tx
```

Comment:
I don't think this would give us anything. We likely have to add inputs and outputs to make it balanced before we can sign it, so adding the signatures here would only make sense if it was balanced & paid up already. Perhaps we can make this nicer with better types for transactions: https://github.com/input-output-hk/plutus/issues/1320


 (Also the lookup function would have to be `(PubKey -> L.Tx -> Maybe Signature)`)